File: lib/xrc/git.ex
defmodule Xrc.Git do
  @moduledoc """
  Git helpers mirroring the proven flow from start-exercise.sh / end-exercise.sh:
  clean-tree check + `pull --ff-only` before work, and stage/commit/push with a
  `"<track>: complete <exercise>"` message after a successful submit. A
  duplicate-submit guard checks the log for a prior completion commit.
  """

  alias Xrc.Config

  @doc "True if the working tree (tracked files) is clean."
  def clean?(%Config{repo_root: root}) do
    git_ok?(root, ["diff", "--quiet"]) and git_ok?(root, ["diff", "--cached", "--quiet"])
  end

  @doc "Fast-forward pull. Returns :ok | {:error, output}."
  def pull_ff(%Config{repo_root: root}) do
    case git(root, ["pull", "--ff-only"]) do
      {_out, 0} -> :ok
      {out, _} -> {:error, out}
    end
  end

  @doc "True if a prior `\"<track>: complete <exercise>\"` commit exists."
  def already_completed?(%Config{repo_root: root}, track, exercise) do
    {out, _} = git(root, ["log", "--oneline", "--grep=#{track}: complete #{exercise}"])
    String.trim(out) != ""
  end

  @doc """
  Stage the exercise dir, show the stat diff, commit `"<track>: complete <exercise>"`,
  and push. `confirm_fun` is called with the staged stat text and must return a
  boolean. Returns :ok | {:error, reason} | :no_changes | :aborted.
  """
  def commit_and_push(%Config{repo_root: root}, track, exercise, confirm_fun) do
    rel = Path.join(track, exercise)

    if not has_changes?(root, rel) do
      :no_changes
    else
      git(root, ["add", rel <> "/"])
      {stat, _} = git(root, ["diff", "--cached", "--stat"])

      if confirm_fun.(stat) do
        case git(root, ["commit", "-m", "#{track}: complete #{exercise}"]) do
          {_out, 0} ->
            case git(root, ["push"]) do
              {_out, 0} -> :ok
              {out, _} -> {:error, "push failed:\n#{out}"}
            end

          {out, _} ->
            {:error, "commit failed:\n#{out}"}
        end
      else
        git(root, ["reset", "HEAD"])
        :aborted
      end
    end
  end

  defp has_changes?(root, rel) do
    {out, _} = git(root, ["status", "--porcelain", "--", rel])
    String.trim(out) != ""
  end

  defp git_ok?(root, args) do
    {_out, code} = git(root, args)
    code == 0
  end

  defp git(root, args) do
    System.cmd("git", ["-C", root | args], stderr_to_stdout: true)
  end
end

File: internal/exercism/git.go
package exercism

import (
	"fmt"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/AirmanBugs/exercism/xrc/internal/config"
)

// GitClean reports whether the working tree (tracked files) is clean.
func GitClean(cfg *config.Config) bool {
	return gitOK(cfg, "diff", "--quiet") && gitOK(cfg, "diff", "--cached", "--quiet")
}

// GitPullFF fast-forwards to sync progress from other devices. Succeeds (no-op)
// when the branch has no upstream; errors only on a real failure.
func GitPullFF(cfg *config.Config) error {
	if !gitOK(cfg, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}") {
		return nil
	}
	if out, _, err := git(cfg, "pull", "--ff-only"); err != nil {
		return fmt.Errorf("%s", strings.TrimSpace(out))
	}
	return nil
}

// AlreadyCompleted reports whether a prior "<track>: complete <exercise>" commit exists.
func AlreadyCompleted(cfg *config.Config, track, exercise string) bool {
	out, _, _ := git(cfg, "log", "--oneline", "--grep="+track+": complete "+exercise)
	return strings.TrimSpace(out) != ""
}

// CommitResult is the outcome of CommitAndPush.
type CommitResult int

const (
	CommitOK CommitResult = iota
	CommitNoChanges
	CommitAborted
)

// CommitAndPush stages the exercise dir, shows the stat diff via confirm, then
// commits "<track>: complete <exercise>" and pushes. confirm receives the staged
// stat and returns whether to proceed.
func CommitAndPush(cfg *config.Config, track, exercise string, confirm func(stat string) bool) (CommitResult, error) {
	rel := filepath.Join(track, exercise)

	if !gitHasChanges(cfg, rel) {
		return CommitNoChanges, nil
	}

	if _, _, err := git(cfg, "add", rel+"/"); err != nil {
		return CommitOK, err
	}
	stat, _, _ := git(cfg, "diff", "--cached", "--stat")

	if !confirm(stat) {
		_, _, _ = git(cfg, "reset", "HEAD")
		return CommitAborted, nil
	}

	if out, _, err := git(cfg, "commit", "-m", track+": complete "+exercise); err != nil {
		return CommitOK, fmt.Errorf("commit failed:\n%s", out)
	}
	if out, _, err := git(cfg, "push"); err != nil {
		return CommitOK, fmt.Errorf("push failed:\n%s", out)
	}
	return CommitOK, nil
}

func gitHasChanges(cfg *config.Config, rel string) bool {
	out, _, _ := git(cfg, "status", "--porcelain", "--", rel)
	return strings.TrimSpace(out) != ""
}

func gitOK(cfg *config.Config, args ...string) bool {
	_, code, _ := git(cfg, args...)
	return code == 0
}

// git runs a git command in the repo root and returns combined output, exit code, error.
func git(cfg *config.Config, args ...string) (string, int, error) {
	full := append([]string{"-C", cfg.RepoRoot}, args...)
	cmd := exec.Command("git", full...)
	out, err := cmd.CombinedOutput()
	code := cmd.ProcessState.ExitCode()
	return string(out), code, err
}
